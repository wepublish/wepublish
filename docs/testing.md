# Automated Testing

Automated testing is a crucial aspect of software development that helps to ensure the quality of a codebase by
automating the testing process.

## API Unit Testing

API unit testing is a type of automated testing that focuses on testing individual units of code that make up an API.

### Setting up the Testing Environment

To set up the testing environment, you need to ensure that you have the following entry in your `jest.config.ts` file: `globalSetup: '<rootDir>/setup-database.js'`. You can find an example of this in the `libs/membership/api/src` folder.

In the `setup-database.js` file, you should have code that looks like the following:

```typescript
import execa from 'execa'

export default async () => {
    const databaseUrl = 'postgresql://postgres@localhost:5432/wepublish_test?schema=public'
    process.env.DATABASE_URL = databaseUrl
    process.env.TZ = 'UTC'

    await execa(`npx`, ['prisma', 'migrate', 'reset', '--force'])
}
```
This code sets up a test database called `wepublish_test` and resets it each time you run a test.

### Running Tests
To run tests, you need to ensure that all your changes to the `schema.prisma` file are reflected in a migration. 
You can do this by running `prisma migrate dev --name name_of_my_migration`. Otherwise, your test database may 
rely on an incomplete schema.
Also, make sure, you have a database up and running. You may want run `docker-compose up database`

Once you have done this, you should be able to run a single test using the command `npx jest -t 'my test'`

### Writing Tests with Fabbrica
When it comes to writing tests, we highly recommend using [Fabbrica](https://github.com/Quramy/prisma-fabbrica).
With Fabbrica, you can easily generate database entries using the `.create()` method or model instances using
the `.build()` method, saving you time and effort.

For examples of how to use Fabbrica in your tests, check out the `subscription-flow.controller.spec.ts` file 
located at `libs/membership/api/src/lib/subscription-flow`. It's a great resource to help you get started with writing
tests that utilize Fabbrica.

### ! Do not forget adding tests to the pipeline !
In order to run your tests within the pipeline, you have to add an entry in the `package.json` in the root folder.
To test, if it works, run `nx test` from your cli. Your specs should now be tested.


## Snapshots
It might happen that your tests are failing. Try to update your snapshots
by running (eventually in the package in question) `npm run test -- -u`

For more information: [https://jestjs.io/docs/cli#--updatesnapshot](https://jestjs.io/docs/cli#--updatesnapshot)