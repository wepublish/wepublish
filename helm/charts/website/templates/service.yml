---
apiVersion: v1
kind: Service
metadata:
  name: {{ include "helm.fullname" . }}
  labels:
    app.kubernetes.io/name: {{ include "helm.fullname" . }}
    helm.sh/chart: {{ include "helm.chart" . }}
    app.kubernetes.io/instance: {{ .Release.Name }}
    app.kubernetes.io/managed-by: {{ .Release.Service }}
    {{- with .Values.labels }}
      {{- toYaml . | nindent 4 }}
    {{- end }}
spec:
  type: {{ .Values.service.type }}
  ports:
    - port: {{ .Values.service.port }}
      targetPort: http
      protocol: TCP
      name: http
  selector:
    app.kubernetes.io/name: {{ include "helm.name" . }}
    app.kubernetes.io/instance: {{ .Release.Name }}
