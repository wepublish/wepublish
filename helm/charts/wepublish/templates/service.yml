---
apiVersion: v1
kind: Service
metadata:
  name: {{ include "helm.fullname" . }}-api
  labels:
    app.kubernetes.io/name: {{ include "helm.fullname" . }}-api
    helm.sh/chart: {{ include "helm.chart" . }}
    app.kubernetes.io/instance: {{ .Release.Name }}
    app.kubernetes.io/managed-by: {{ .Release.Service }}
    {{- with .Values.labels }}
      {{- toYaml . | nindent 4 }}
    {{- end }}
spec:
  type: {{ .Values.service.api.type }}
  ports:
    - port: {{ .Values.service.api.port }}
      targetPort: http
      protocol: TCP
      name: http
  selector:
    app.kubernetes.io/name: {{ include "helm.name" . }}-api
    app.kubernetes.io/instance: {{ .Release.Name }}
---
apiVersion: v1
kind: Service
metadata:
  name: {{ include "helm.fullname" . }}-editor
  labels:
    app.kubernetes.io/name: {{ include "helm.fullname" . }}-editor
    helm.sh/chart: {{ include "helm.chart" . }}
    app.kubernetes.io/instance: {{ .Release.Name }}
    app.kubernetes.io/managed-by: {{ .Release.Service }}
    {{- with .Values.labels }}
    {{- toYaml . | nindent 4 }}
    {{- end }}
spec:
  type: {{ .Values.service.editor.type }}
  ports:
    - port: {{ .Values.service.editor.port }}
      targetPort: http
      protocol: TCP
      name: http
  selector:
    app.kubernetes.io/name: {{ include "helm.name" . }}-editor
    app.kubernetes.io/instance: {{ .Release.Name }}
