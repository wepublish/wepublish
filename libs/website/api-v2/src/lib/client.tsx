import {
  ApolloClient,
  InMemoryCache,
  ApolloProvider,
  ApolloLink,
  NormalizedCacheObject
} from '@apollo/client'
import {ComponentType, memo, useMemo} from 'react'
import {BatchHttpLink} from '@apollo/client/link/batch-http'

export const getV2ApiClient = (
  apiUrl: string,
  links: ApolloLink[],
  cache?: NormalizedCacheObject
) => {
  const httpLink = new BatchHttpLink({
    uri: `${apiUrl}/v2`,
    batchMax: 5,
    batchInterval: 20
  })
  const link = [...links, httpLink].reduce(
    (links: ApolloLink | undefined, link) => (links ? links.concat(link) : link),
    undefined
  )

  return new ApolloClient({
    link,
    cache: new InMemoryCache().restore(cache ?? {}),
    ssrMode: typeof window === 'undefined'
  })
}

export const useV2ApiClient = (
  apiUrl: string,
  links: ApolloLink[] = [],
  cache?: NormalizedCacheObject
) => useMemo(() => getV2ApiClient(apiUrl, links, cache), [apiUrl, links, cache])

export const createWithV2ApiClient =
  (apiUrl: string, links: ApolloLink[] = [], cache?: NormalizedCacheObject) =>
  <
    // eslint-disable-next-line @typescript-eslint/ban-types
    P extends object
  >(
    ControlledComponent: ComponentType<P>
  ) =>
    memo<P>(props => {
      const client = useV2ApiClient(apiUrl, links, cache)

      return (
        <ApolloProvider client={client}>
          <ControlledComponent {...props} />
        </ApolloProvider>
      )
    })
