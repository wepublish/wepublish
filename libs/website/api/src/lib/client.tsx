import {
  ApolloClient,
  InMemoryCache,
  ApolloProvider,
  ApolloLink,
  NormalizedCacheObject
} from '@apollo/client'
import {BatchHttpLink} from '@apollo/client/link/batch-http'

import {ComponentType, memo, useMemo} from 'react'

export const getV1ApiClient = (
  apiUrl: string,
  links: ApolloLink[],
  cache?: NormalizedCacheObject
) => {
  const httpLink = new BatchHttpLink({uri: `${apiUrl}/v1`, batchMax: 5, batchInterval: 20})
  const link = [...links, httpLink].reduce(
    (links: ApolloLink | undefined, link) => (links ? links.concat(link) : link),
    undefined
  )

  return new ApolloClient({
    link,
    cache: new InMemoryCache().restore(cache ?? {}),
    ssrMode: typeof window === 'undefined'
  })
}

export const useV1ApiClient = (
  apiUrl: string,
  links: ApolloLink[] = [],
  cache?: NormalizedCacheObject
) => useMemo(() => getV1ApiClient(apiUrl, links, cache), [apiUrl, links, cache])

export const createWithV1ApiClient =
  (apiUrl: string, links: ApolloLink[] = [], cache?: NormalizedCacheObject) =>
  <
    // eslint-disable-next-line @typescript-eslint/ban-types
    P extends object
  >(
    ControlledComponent: ComponentType<P>
  ) =>
    memo<P>(props => {
      const client = useV1ApiClient(apiUrl, links, cache)

      return (
        <ApolloProvider client={client}>
          <ControlledComponent {...props} />
        </ApolloProvider>
      )
    })
