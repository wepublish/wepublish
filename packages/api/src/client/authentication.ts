import {query, RequestOptions} from './query'

export async function authenticateWithCredentials(
  url: string,
  email: string,
  password: string,
  opts?: RequestOptions
): Promise<{
  user: {email: string}
  refreshToken: string
  accessToken: string
  refreshTokenExpiresIn: number
  accessTokenExpiresIn: number
}> {
  const {authenticateWithCredentials: response} = await query(
    url,
    {
      query: `
    mutation authenticateWithCredentials($email: String!, $password: String!) {
      authenticateWithCredentials(email: $email, password: $password) {
        user {email},
        refreshToken,
        accessToken,
        refreshTokenExpiresIn,
        accessTokenExpiresIn
      }
    }
  `,
      variables: {
        email,
        password
      }
    },
    opts
  )
  return response
}
