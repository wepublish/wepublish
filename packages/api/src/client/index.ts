export * from './types'
export * from './utility'
export * from './queryArticles'
