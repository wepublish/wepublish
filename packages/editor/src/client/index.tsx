import 'core-js/stable'
import 'regenerator-runtime/runtime'

import React from 'react'
import ReactDOM from 'react-dom'
import {ApolloProvider, ApolloClient, ApolloLink, InMemoryCache} from '@apollo/client'
import {createUploadLink} from 'apollo-upload-client'
import {onError} from '@apollo/client/link/error'

import './i18n'

import {ElementID} from '../shared/elementID'
import {ClientSettings} from '../shared/types'
import {RouteProvider} from './route'
import {AuthProvider} from './authContext'
import {LocalStorageKey} from './utility'
import {TwitterProvider} from './blocks/embeds/twitter'
import {InstagramProvider} from './blocks/embeds/instagram'
import {FacebookProvider} from './blocks/embeds/facebook'
import {HotApp} from './app'

// See: https://www.apollographql.com/docs/react/data/fragments/#fragments-on-unions-and-interfaces
export async function fetchIntrospectionQueryResultData(url: string) {
  const response = await fetch(url, {
    method: 'POST',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify({
      variables: {},
      query: `
      {
        __schema {
          types {
            kind
            name
            possibleTypes {
              name
            }
          }
        }
      }
    `
    })
  })

  const result = await response.json()

  const possibleTypes: any = {}

  result.data.__schema.types.forEach((supertype: any) => {
    if (supertype.possibleTypes) {
      possibleTypes[supertype.name] = supertype.possibleTypes.map((subtype: any) => subtype.name)
    }
  })

  return possibleTypes
}

const onDOMContentLoaded = async () => {
  const {apiURL}: ClientSettings = JSON.parse(
    document.getElementById(ElementID.Settings)!.textContent!
  )

  const adminAPIURL = `${apiURL}/admin`

  const authLink = new ApolloLink((operation, forward) => {
    const token = localStorage.getItem(LocalStorageKey.SessionToken)
    const context = operation.getContext()

    operation.setContext({
      headers: {
        authorization: token ? `Bearer ${token}` : '',
        ...context.headers
      },
      ...context
    })

    return forward(operation)
  })

  const authErrorLink = onError(({graphQLErrors, /* networkError, */ operation, forward}) => {
    if (graphQLErrors) {
      graphQLErrors.forEach(({/* message, locations, path, */ extensions}) => {
        // console.log(`[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`)
        if (extensions?.code === 'UNAUTHENTICATED') {
          window.location.pathname = '/logout'
        }
      })
      // if (networkError) console.log(`[Network error]: ${networkError}`)
    }
    return forward(operation)
  })

  const client = new ApolloClient({
    link: authLink.concat(authErrorLink).concat(createUploadLink({uri: adminAPIURL})),
    cache: new InMemoryCache({
      possibleTypes: await fetchIntrospectionQueryResultData(adminAPIURL)
    })
  })

  window.addEventListener('dragover', e => e.preventDefault())
  window.addEventListener('drop', e => e.preventDefault())

  ReactDOM.render(
    <ApolloProvider client={client}>
      <AuthProvider>
        <RouteProvider>
          <FacebookProvider sdkLanguage={'en_US'}>
            <InstagramProvider>
              <TwitterProvider>
                <HotApp />
              </TwitterProvider>
            </InstagramProvider>
          </FacebookProvider>
        </RouteProvider>
      </AuthProvider>
    </ApolloProvider>,
    document.getElementById(ElementID.ReactRoot)
  )
}

if (document.readyState !== 'loading') {
  onDOMContentLoaded()
} else {
  document.addEventListener('DOMContentLoaded', onDOMContentLoaded)
}
