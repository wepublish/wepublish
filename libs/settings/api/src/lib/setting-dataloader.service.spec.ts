import {Test, TestingModule} from '@nestjs/testing'
import {SettingDataloaderService} from './setting-dataloader.service'
import {PrismaClient} from '@prisma/client'
import DataLoader from 'dataloader'

jest.mock('dataloader')

describe('SettingDataloaderService', () => {
  let service: SettingDataloaderService
  let prismaMock: {
    setting: {
      findMany: jest.Mock
    }
  }

  beforeEach(async () => {
    prismaMock = {
      setting: {
        findMany: jest.fn()
      }
    }

    const module: TestingModule = await Test.createTestingModule({
      providers: [
        SettingDataloaderService,
        {
          provide: PrismaClient,
          useValue: prismaMock
        }
      ]
    }).compile()

    service = await module.resolve<SettingDataloaderService>(SettingDataloaderService)
  })

  it('should prime', () => {
    // @ts-expect-error Mock so typings incorrectly
    const dataloaderMock = DataLoader.mock.instances[0] as jest.fn
    service.prime('123', {} as any)
    expect(dataloaderMock.prime.mock.calls[0]).toMatchSnapshot()
  })

  describe('load', () => {
    beforeEach(async () => {
      // @ts-expect-error mocked so typing doesn't work
      DataLoader.mockImplementation((impl, opt) => {
        return {
          load: (id: string) => impl([id]),
          loadMany: (ids: readonly string[]) => impl(ids)
        }
      })

      const module: TestingModule = await Test.createTestingModule({
        providers: [
          SettingDataloaderService,
          {
            provide: PrismaClient,
            useValue: prismaMock
          }
        ]
      }).compile()

      service = await module.resolve<SettingDataloaderService>(SettingDataloaderService)
    })

    it('should load one', async () => {
      prismaMock.setting.findMany.mockResolvedValue([])

      await service.load('123')
      expect(prismaMock.setting.findMany).toHaveBeenCalled()
      expect(prismaMock.setting.findMany.mock.calls[0]).toMatchSnapshot()
    })

    it('should load many', async () => {
      prismaMock.setting.findMany.mockResolvedValue([])

      await service.loadMany(['123', '321'])
      expect(prismaMock.setting.findMany).toHaveBeenCalled()
      expect(prismaMock.setting.findMany.mock.calls[0]).toMatchSnapshot()
    })
  })
})
