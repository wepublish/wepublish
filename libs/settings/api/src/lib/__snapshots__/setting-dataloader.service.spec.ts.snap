// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingDataloaderService load should load many 1`] = `
[
  {
    "where": {
      "name": {
        "in": [
          "123",
          "321",
        ],
      },
    },
  },
]
`;

exports[`SettingDataloaderService load should load one 1`] = `
[
  {
    "where": {
      "name": {
        "in": [
          "123",
        ],
      },
    },
  },
]
`;

exports[`SettingDataloaderService should prime 1`] = `
[
  "123",
  {},
]
`;
